Source/WTF/wtf/text/StringConcatenate.h

```diff
@@ -458,16 +458,140 @@ template<> class StringTypeAdapter<ASCIICaseConverter, void> {
+            accumulator.append(each(*begin));
+
+            ++begin;
+            for (; begin != end; ++begin)
+                accumulator.append(between, each(*begin));
+        }
+    }
+};
+
+// The `interleave` function can be called in three different ways:
+//
+//  1. The most generic way provides an `each` functor taking two arguments,
+//     the `accumulator` and the `value`, and returns `void`.
+//
+//       Vector<Foo> container = { ... };
+//
+//       ... interleave(
+//              container,
+//              [](auto& accumulator, auto& value) {
+//                  accumulator.append(value.stringRepresentation(), '-', value.otherStringRepresentation());
+//              },
+//              ", "_s
+//           ), ...
+//
+//     This allows for containers of non-string values to provide complex mapped
+//     values without additional allocations.
+//
+//  2. If multiple mapped strings per-value are not required, an `each` functor
+//     taking just the `value` and returning a "string-type" (i.e. something you
+//     could pass to `StringBuilder::append(...)`).
+//
+//       Vector<Foo> container = { ... };
+//
+//       ... interleave(
+//              container,
+//              [](auto& value) {
+//                  return value.stringRepresentation();
+//              },
+//              ", "_s
+//           ), ...
+//
+//  3. Finally, if the container already contains "string-types", no `each` functor
+//     is required at all.
+//
+//       Vector<String> container = { ... };
+//
+//       ... interleave(
+//              container,
+//              ", "_s
+//           ), ...
+//
+template<typename C, typename E, typename B>
+decltype(auto) interleave(C&& container, E&& each, B&& between)
+{
+    return Interleave {
+        std::forward<C>(container),
+        std::forward<E>(each),
+        std::forward<B>(between)
+    };
+}
+
+template<typename C, typename B>
+decltype(auto) interleave(C&& container, B&& between)
+{
+    return Interleave {
+        std::forward<C>(container),
+        [](auto& accumulator, auto& value) { accumulator.append(value); },
+        std::forward<B>(between)
+    };
+}
+
+template<typename C, typename E, typename B>
+class StringTypeAdapter<Interleave<C, E, B>, void> {
+public:
+    StringTypeAdapter(const Interleave<C, E, B>& interleave)
+        : m_interleave { interleave }
+    {
+    }
+
+    template<typename Accumulator>
+    void writeUsing(Accumulator&& accumulator) const
+    {
+        m_interleave.writeUsing(std::forward<Accumulator>(accumulator));
+    }
+
+private:
+    const Interleave<C, E, B>& m_interleave;
+};
+
+template<typename... Adapters>
+inline bool are8Bit(Adapters&& ...adapters)
+{
+    return (... && adapters.is8Bit());
+}
+
 template<typename Adapter>
-inline bool are8Bit(Adapter adapter)
+constexpr bool hasKnownLength(Adapter&&)
 {
-    return adapter.is8Bit();
+    return requires(Adapter x) {
+        { x.length() } -> std::same_as<unsigned>;
```

Comment:
Could this subtly fail if `length()` returns `size_t` (or some other integer type) and cause us to take the slow path? And if so, how can we be sure to notice this if we make the error?

Reply:
Ok, I have two ideas:

1. Change std::same_as<unsigned> to `std::integral`, since I think that would cover all reasonable cases.

but I think it would be safe to error the other way and switch this to a check for the `writeUsing()` function that only the "slow" version has.

Reply:
Either of those sounds good!